File: spy/vm/bytes.py
from typing import TYPE_CHECKING

from spy.fqn import FQN
from spy.libspy import LLSPyInstance
from spy.vm.b import BUILTINS, B
from spy.vm.builtin import builtin_method
from spy.vm.object import W_Object
from spy.vm.primitive import W_I32

if TYPE_CHECKING:
    from spy.vm.vm import SPyVM


def ll_bytes_new(ll: LLSPyInstance, b: bytes) -> int:
    """
    Create a new spy_BytesObject inside the given LLSPyInstance, filling it
    with the content of b.

    Returns the corresponding 'spy_BytesObject *'
    """
    length = len(b)
    ptr = ll.call("spy_bytes_alloc", length)
    data_ptr = ll.mem.read_i32(ptr + ll.bytes_layout.data_offset)
    ll.mem.write(data_ptr, b)
    return ptr


@B.builtin_type("bytes", lazy_definition=True)
class W_Bytes(W_Object):
    """
    A sequence of bytes.

    This is basically a 'spy_BytesObject *', i.e. a pointer to a C struct
    which resides in the linear memory of the VM:
        typedef struct {
            size_t length;
            int32_t hash;
            uint8_t *data;
        } spy_BytesObject;
    """

    __spy_storage_category__ = "value"
    __spy_lazy_attributes__ = {
        "__len__": FQN("_bytes::methods::__len__"),
        "__getitem__": FQN("_bytes::methods::__getitem__"),
        "__eq__": FQN("_bytes::methods::__eq__"),
        "__ne__": FQN("_bytes::methods::__ne__"),
        "__add__": FQN("_bytes::methods::__add__"),
        "__mul__": FQN("_bytes::methods::__mul__"),
        "__repr__": FQN("_bytes::methods::__repr__"),
    }

    vm: "SPyVM"
    ptr: int

    def __init__(self, vm: "SPyVM", b: bytes) -> None:
        self.vm = vm
        self.ptr = ll_bytes_new(vm.ll, b)

    @staticmethod
    def from_ptr(vm: "SPyVM", ptr: int) -> "W_Bytes":
        w_res = W_Bytes.__new__(W_Bytes)
        w_res.vm = vm
        w_res.ptr = ptr
        return w_res

    def get_length(self) -> int:
        ll = self.vm.ll
        return ll.mem.read_i32(self.ptr + ll.bytes_layout.length_offset)

    def get_data(self) -> bytes:
        _, _, data = self.vm.ll.read_bytes(self.ptr)
        return data

    def __repr__(self) -> str:
        b = self.get_data()
        return f"W_Bytes({b!r})"

    def spy_unwrap(self, vm: "SPyVM") -> bytes:
        return self.get_data()

    def spy_key(self, vm: "SPyVM") -> bytes:
        return self.get_data()

    # XXX: __hash__ stays in C because applevel SPy has no wrapping i32
    # multiply yet (needed for FNV-1a). Once wrapping arithmetic is
    # available, this can move to _bytes.spy methods like the other dunders.
    @builtin_method("__hash__")
    @staticmethod
    def w_hash(vm: "SPyVM", w_b: "W_Bytes") -> W_I32:
        assert isinstance(w_b, W_Bytes)
        result = vm.ll.call("spy_bytes_hash", w_b.ptr)
        return vm.wrap(result)

File: spy/libspy/src/bytes.c
#include "spy.h"
#include <stddef.h>
#include <stdint.h>

_spy_BytesObject_Layout
_spy_BytesObject_layout(void) {
    return (_spy_BytesObject_Layout){
        .size = sizeof(spy_BytesObject),
        .length_offset = offsetof(spy_BytesObject, length),
        .hash_offset = offsetof(spy_BytesObject, hash),
        .data_offset = offsetof(spy_BytesObject, data),
    };
}

spy_BytesObject *
spy_bytes_alloc(size_t length) {
    // allocate a spy_BytesObject AND the data buffer as a single allocation
    size_t size = sizeof(spy_BytesObject) + length;
    spy_BytesObject *res = (spy_BytesObject *)spy_GcAlloc(size).p;
    res->length = length;
    res->hash = 0;
#ifdef SPY_DEBUG
    res->data = (spy_gc_ptr_u8){(uint8_t *)(res + 1), (ptrdiff_t)length};
#else
    res->data = (spy_gc_ptr_u8){(uint8_t *)(res + 1)};
#endif
    return res;
}

int32_t
spy_bytes_hash(spy_BytesObject *b) {
    if (b->hash != 0)
        return b->hash;
    // FNV-1a hash (same algorithm as str; hashes not required to match)
    // NOTE: must stay in C because applevel SPy has no wrapping i32 multiply yet
    uint32_t h = 2166136261u;
    for (size_t i = 0; i < b->length; i++) {
        h ^= spy_BytesObject_DATA(b)[i];
        h *= 16777619u;
    }
    int32_t result = (int32_t)h;
    if (result == -1)
        result = -2;
    if (result == 0)
        result = 1;
    b->hash = result;
    return result;
}
